src/storage/caching_file_system.cpp

```diff
@@ -0,0 +1,415 @@
+}
+
+CachingFileHandle::CachingFileHandle(CachingFileSystem &caching_file_system_p, CachedFile &cached_file_p,
+                                     FileOpenFlags flags_p)
+    : caching_file_system(caching_file_system_p), cached_file(cached_file_p), flags(flags_p) {
+	if (!caching_file_system.enable || caching_file_system.check_cached_file_invalidation) {
+		// If caching is disabled, or if we must check cache invalidation, we always have to open the file
+		GetFileHandle();
+		return;
+	}
+	// If we don't have any cached file ranges, we must also open the file
+	auto guard = cached_file.lock.GetSharedLock();
+	if (cached_file.Ranges(guard).empty()) {
+		GetFileHandle();
+	}
+}
+
+FileHandle &CachingFileHandle::GetFileHandle() {
+	if (!file_handle) {
+		const auto current_time = duration_cast<std::chrono::seconds>(system_clock::now().time_since_epoch()).count();
+		file_handle = caching_file_system.file_system.OpenFile(cached_file.path, flags);
+		last_modified = caching_file_system.file_system.GetLastModifiedTime(*file_handle);
+		version_tag = caching_file_system.file_system.GetVersionTag(*file_handle);
+
+		auto guard = cached_file.lock.GetExclusiveLock();
+		if (!caching_file_system.FileIsValid(cached_file, guard, version_tag, last_modified, current_time)) {
+			cached_file.Ranges(guard).clear(); // Invalidate entire cache
+		}
+		cached_file.FileSize(guard) = file_handle->GetFileSize();
+		cached_file.LastModified(guard) = last_modified;
+		cached_file.VersionTag(guard) = version_tag;
+		cached_file.CanSeek(guard) = file_handle->CanSeek();
+		cached_file.OnDiskFile(guard) = file_handle->OnDiskFile();
+	}
+	return *file_handle;
+}
+
+BufferHandle CachingFileHandle::Read(data_ptr_t &buffer, const idx_t nr_bytes, const idx_t location) {
+	BufferHandle result;
+	if (!caching_file_system.enable) {
+		result = caching_file_system.buffer_manager.Allocate(MemoryTag::EXTERNAL_FILE_CACHE, nr_bytes);
+		buffer = result.Ptr();
+		GetFileHandle().Read(buffer, nr_bytes, location);
+		return result;
+	}
+
+	// Get read lock for cached ranges
+	auto guard = cached_file.lock.GetSharedLock();
+	auto &ranges = cached_file.Ranges(guard);
+
+	// First, try to see if we've read from the exact same location before
+	auto it = ranges.find(location);
+	if (it != ranges.end()) {
+		// We have read from the exact same location before
+		if (it->second->GetOverlap(nr_bytes, location) == CachedFileRangeOverlap::FULL) {
+			// The file range contains the requested file range
+			result = TryReadFromFileRange(guard, *it->second, buffer, nr_bytes, location);
+			if (result.IsValid()) {
+				return result;
+			}
+		}
+	}
+
+	// Second, loop through file ranges (ordered by location) to see if any contain the requested file range
+	const auto this_end = location + nr_bytes;
+	vector<shared_ptr<CachedFileRange>> overlapping_ranges;
+	for (it = ranges.begin(); it != ranges.end();) {
```

Comment:
We shouldn't need to start at `ranges.begin()` here - we can finding the range right before our starting location and start looping there